File: practica_03/practica_glm_auto_student.py
"""
Práctica 3 – Aplicación de GLMs en Pricing de Autos (Versión para estudiantes)

Objetivo
--------
Analizar la frecuencia y severidad de siniestros en un portafolio de autos
utilizando Modelos Lineales Generalizados (GLMs). Al final calcularás la prima
pura por segmentos. Esta plantilla contiene las secciones que debes completar.

Datos
-----
Utiliza los conjuntos `freMTPL2freq.csv` y `freMTPL2sev.csv` (French Motor
Third‑Party Liability). El primero contiene las variables de riesgo y el número
de siniestros (`ClaimNb`) junto con la exposición (`Exposure`); el segundo
contiene el monto de reclamaciones (`ClaimAmount`)【26693141931704†L25-L43】. Coloca ambos
archivos en la carpeta `data/`.

Instrucciones
-------------
- Completa cada sección marcada con `TODO`. Utiliza bibliotecas como
  `pandas`, `numpy` y `statsmodels` para los modelos GLM.
- Comenta tu código y explica cada paso brevemente.
- Al final prepara un reporte breve con tus conclusiones.
"""

import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf

# 1. Lectura y unión de datos
# -----------------------------------------------------------------------------
# TODO: Carga los datos freMTPL2freq y freMTPL2sev desde archivos CSV en la
# carpeta `data/`. Utiliza `pd.read_csv()` y especifica el separador correcto.
# A continuación, agrupa los montos de reclamaciones por `IDpol` en el dataset
# de severidad y une ambos dataframes usando `merge()` en el identificador.

# Ejemplo (descomenta y ajusta las rutas):
# freq = pd.read_csv('data/freMTPL2freq.csv')
# sev = pd.read_csv('data/freMTPL2sev.csv')
# sev_agg = sev.groupby('IDpol', as_index=False)['ClaimAmount'].sum()
# df = freq.merge(sev_agg, how='left', on='IDpol')
# df['ClaimAmount'] = df['ClaimAmount'].fillna(0)

# 2. Preparación de variables
# -----------------------------------------------------------------------------
# TODO:
# - Calcula la frecuencia observada: Frequency = ClaimNb / Exposure.
# - Calcula la severidad observada: Severity = ClaimAmount / max(ClaimNb, 1).
# - Convierte las variables categóricas a tipo `category` utilizando
#   `df['Area'] = df['Area'].astype('category')`, etc.
# - Explora el dataset con `df.info()` y revisa valores faltantes.

# 3. Modelo de frecuencia
# -----------------------------------------------------------------------------
# 3a. Ajusta un GLM Poisson con offset
# Usa `statsmodels` para ajustar un modelo Poisson con enlace log. El offset
# es `np.log(Exposure)` y se pasa como argumento en `sm.GLM`. Selecciona
# algunas variables explicativas (por ejemplo: Area, VehPower, VehAge, DrivAge,
# BonusMalus). Observa el resumen con `result.summary()`.

# TODO: escribe el código para ajustar el modelo Poisson. Ejemplo de fórmula:
# formula = 'ClaimNb ~ C(Area) + C(VehPower) + VehAge + DrivAge + BonusMalus'
# model_pois = smf.glm(formula=formula, data=df,
#                      family=sm.families.Poisson(),
#                      offset=np.log(df['Exposure']))
# result_pois = model_pois.fit()
# print(result_pois.summary())

# Calcula media y varianza de la frecuencia para evaluar sobredispersión, y el
# ratio deviance/df (`result_pois.deviance / result_pois.df_resid`)【718163142841177†L702-L709】.

# 3b. Ajusta un modelo de Binomial Negativa si detectas sobredispersión
# Usa `statsmodels` con `sm.families.NegativeBinomial()`. Puedes fijar
# `alpha=None` para que el software estime el parámetro de dispersión.

# TODO: ajusta el modelo NB y compara sus métricas (AIC, deviance) con el modelo
# Poisson. Interpreta los coeficientes usando `np.exp(coef) - 1`.

# 4. Modelo de severidad
# -----------------------------------------------------------------------------
# Filtra las pólizas con `ClaimAmount > 0` porque la distribución Gamma tiene
# soporte en valores positivos【718163142841177†L850-L852】.
# Ajusta un GLM Gamma con enlace log usando `statsmodels`:
# formula_sev = 'Severity ~ C(Area) + C(VehPower) + VehAge + DrivAge + BonusMalus'
# model_gamma = smf.glm(formula=formula_sev, data=df_sev,
#                       family=sm.families.Gamma(link=sm.genmod.families.links.log()),
#                       weights=df_sev['ClaimNb'])
# result_gamma = model_gamma.fit()
# print(result_gamma.summary())

# 4b. Exclusión de siniestros grandes
# Calcula un percentil alto (por ejemplo 95%) de `ClaimAmount` y filtra las
# observaciones por debajo de ese umbral【128841580829174†L4018-L4028】. Ajusta
# nuevamente el modelo Gamma y observa cambios en los coeficientes.

# 5. Cálculo de la prima pura
# -----------------------------------------------------------------------------
# Obtén predicciones de frecuencia y severidad con `result_pois.predict()` o
# `result_nb.predict()` y `result_gamma.predict()`. Multiplica ambas
# predicciones para obtener la prima pura: `pure_premium = freq_pred * sev_pred`
#【718163142841177†L653-L662】.

# TODO: genera `df['PurePremium_pred']` e interpreta los valores.

# 6. Segmentación y comparación de primas
# -----------------------------------------------------------------------------
# Agrupa las primas puras predichas por segmentos relevantes (p.ej. bandas de
# edad del conductor, región o tipo de área) utilizando `groupby()` y
# `agg(['mean','count'])`. Crea una tabla comparativa y, si lo deseas,
# grafica los resultados usando `matplotlib` o `seaborn`.

# TODO: construye la tabla de comparación de primas puras por segmento.

# 7. Reporte
# -----------------------------------------------------------------------------
# Prepara un breve reporte (máx. 2 páginas) respondiendo:
# 1. ¿Qué modelo ajusta mejor la frecuencia (Poisson o Neg. Binomial)?
# 2. ¿Qué variables son más relevantes para la severidad?
# 3. ¿Cómo se comparan las primas puras entre segmentos (e.g. jóvenes vs. adultos,
#    urbano vs. rural, uso particular vs. comercial)?

# Incluye tablas, métricas de ajuste y gráficos para justificar tus respuestas.

File: practica_03/practica_glm_auto_student.R
# Práctica 3 – Aplicación de GLMs en Pricing de Autos (Versión para estudiantes)
#
# Objetivo
# --------
# En esta práctica analizarás la frecuencia y severidad de siniestros en un
# portafolio de autos para estimar la prima pura por segmentos. Utilizarás
# Modelos Lineales Generalizados (GLMs) y evaluarás alternativas en caso de
# sobredispersión. Al finalizar, deberás calcular la prima pura esperada y
# elaborar un breve reporte de conclusiones.

# Datos
# -----
# Se utilizarán los conjuntos de datos `freMTPL2freq` y `freMTPL2sev` que
# pertenecen al **French Motor Third‑Party Liability**. El primero contiene
# las características de riesgo y el número de siniestros (`ClaimNb`) junto con
# la exposición (`Exposure`), mientras que el segundo contiene el monto de
# reclamación (`ClaimAmount`) para cada póliza【26693141931704†L25-L43】. Debes
# descargar ambos archivos en formato CSV y guardarlos en una carpeta `data/`.

# 1. Carga de paquetes -------------------------------------------------------
library(tidyverse)   # manipulación y gráficos
library(MASS)        # para glm.nb (Binomial Negativa)
library(broom)       # para resumir modelos

# 2. Lectura y unión de datos -----------------------------------------------

# TODO: lee los archivos CSV y crea un data frame para cada uno. Por ejemplo:
# freq <- read_csv("data/freMTPL2freq.csv")
# sev  <- read_csv("data/freMTPL2sev.csv")

# TODO: combina las tablas utilizando el campo IDpol/PolicyID (ajusta el nombre
# según la versión del archivo). Considera que una póliza puede tener varias
# reclamaciones en sev, por lo que al unir necesitarás sumar los montos de
# reclamación por póliza. Puedes utilizar la función `group_by()` y `summarise()`
# para agregarlos antes de la unión.

# 3. Preparación de variables -----------------------------------------------

# TODO:
# - Crea la variable de frecuencia: frecuencia = ClaimNb / Exposure.
# - Crea la variable de severidad: severidad = ClaimAmount / pmax(ClaimNb, 1).
# - Convierte a factores las variables categóricas como `Area`, `VehPower`,
#   `VehBrand`, `VehGas`, `Region`, `BonusMalus`, etc.
# - Inspecciona el resumen de datos con `glimpse()` y revisa valores faltantes.

# 4. Modelado de la frecuencia ----------------------------------------------

# 4a. Ajuste de un GLM Poisson con offset
# --------------------------------------
# Ajusta un GLM usando distribución Poisson para la variable `ClaimNb` con
# `log(Exposure)` como offset (es decir, le pasa el argumento `offset = log(Exposure)`).
# Selecciona un subconjunto de variables explicativas (por ejemplo: `Area`,
# `VehPower`, `VehAge`, `DrivAge`, `BonusMalus`). Utiliza `glm()` con
# `family = poisson(link="log")` y especifica `weights = Exposure` si deseas
# ponderar por exposición. Revisa el resumen del modelo con `summary()`.

# TODO: escribe el código para ajustar el modelo Poisson.
# Ejemplo de fórmula:
# freq_model <- glm(ClaimNb ~ factor(Area) + factor(VehPower) + VehAge + DrivAge + BonusMalus,
#                   family = poisson(link = "log"), offset = log(Exposure), data = df)

# Calcula la media y la varianza de `ClaimNb/Exposure` y compara para detectar
# posibles signos de sobredispersión. También calcula la razón Deviance/df
# utilizando `deviance(freq_model)/df.residual(freq_model)`【718163142841177†L702-L709】.
# Si la razón es mucho mayor que 1, probablemente exista sobredispersión.

# 4b. Modelo de Binomial Negativa
# -------------------------------
# Si detectas sobredispersión en el modelo Poisson, ajusta un modelo de
# Binomial Negativa utilizando `glm.nb()` del paquete MASS. Usa la misma
# fórmula del modelo Poisson y compara AICs y Deviance/df para determinar cuál
# se ajusta mejor【128841580829174†L4126-L4132】. Interpreta los coeficientes en
# términos de factores multiplicativos (exp(beta)).

# TODO: ajusta el modelo `neg_bin_model <- glm.nb(... )` y extrae un resumen.

# 5. Modelado de la severidad -----------------------------------------------

# La severidad se refiere al costo promedio de los siniestros. Según la
# literatura actuarial, esta variable puede modelarse con una distribución
# Gamma con enlace log【718163142841177†L840-L847】. Debes filtrar solo las
# observaciones con `ClaimAmount > 0` y usar `ClaimNb` como peso para
# reflejar que una póliza puede tener varios reclamos【718163142841177†L850-L866】.

# 5a. Ajuste del modelo Gamma
# ---------------------------
# TODO: ajusta un modelo de severidad con `glm()` usando
# `family = Gamma(link = "log")`. Usa como variable respuesta la severidad
# calculada en la sección 3 y como pesos `ClaimNb`. Incluye las mismas
# variables que en el modelo de frecuencia y revisa el resumen.

# 5b. Exclusión de grandes siniestros
# -----------------------------------
# Los siniestros de gran tamaño pueden distorsionar las estimaciones de
# severidad. Selecciona un umbral alto (por ejemplo el percentil 95 del
# `ClaimAmount`) y excluye las observaciones por encima de dicho umbral【128841580829174†L4018-L4028】.
# Ajusta nuevamente el modelo Gamma y compara los coeficientes y métricas.

# TODO: calcula el percentil 95 y filtra los datos. Ajusta el modelo.

# 5c. Modelo separado para grandes siniestros (opcional)
# -----------------------------------------------------
# Para un análisis más detallado, ajusta un modelo aparte para los siniestros
# que superan el umbral. Puedes explorar una distribución Lognormal utilizando
# `glm()` con familia gaussian y respuesta `log(ClaimAmount)`. Recuerda
# interpretar los coeficientes en términos porcentuales.

# TODO: ajusta un modelo lognormal a las pérdidas grandes (si lo consideras necesario).

# 6. Cálculo de la prima pura -----------------------------------------------

# La prima pura por póliza se obtiene multiplicando la frecuencia estimada
# por la severidad estimada: πᵢ = λ̂ᵢ × μ̂ᵢ【718163142841177†L653-L662】. Una vez
# ajustados los modelos, genera predicciones de frecuencia y severidad para
# cada póliza utilizando `predict()` con `type = "response"`.

# TODO: genera `freq_pred <- predict(freq_model, type = "response")` y
# `sev_pred <- predict(sev_model, type = "response")`. Luego calcula la
# prima pura predicha (`pure_premium <- freq_pred * sev_pred`).

# 7. Segmentación y comparación de primas -----------------------------------

# Agrupa las primas puras predichas por segmentos de interés (p.ej. bandas de
# edad del conductor, región, tipo de área) y calcula el promedio de la
# prima pura en cada grupo. Utiliza `group_by()` y `summarise()` para crear
# una tabla comparativa. Interpreta las diferencias entre segmentos.

# TODO: construye la tabla segmentada y genera gráficos de comparación.

# 8. Reporte
# ----------
# Prepara un documento (máx. 2 páginas) que responda a las preguntas:
# 1. ¿Cuál modelo ajusta mejor la frecuencia (Poisson vs. Binomial Negativa)?
# 2. ¿Qué variables son más relevantes para la severidad?
# 3. ¿Cómo se comparan las primas puras entre segmentos (e.g. jóvenes vs. adultos,
#    urbano vs. rural, uso particular vs. comercial)?

# Incluye tablas de coeficientes, métricas de ajuste (Deviance/df, AIC/BIC) y
# gráficos que apoyen tus respuestas. Interpreta los coeficientes en términos
# de factores multiplicativos (exp(beta) − 1) según la teoría【718163142841177†L702-L709】.

# Recuerda entregar el script completo y el reporte en la fecha indicada.